# Project - Week 4
## Work Plan
### 1. Systematic Study and Practice of Machine Learning

### 2. Application to Thesis: Further Dataset Optimization

### 3. Data Cleaning and Feature Engineering

### 4. Organization and Documentation with Graphics and Annotations


In [79]:
import pandas as pd
df = pd.read_excel("Data_4.xlsx")
selected_columns = ['EmpNumber', 'Age', 'Gender','EducationBackground', 'MaritalStatus', 'EmpDepartment' , 'BusinessTravelFrequency', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpHourlyRate', 'EmpJobInvolvement',  'EmpJobLevel', 'EmpJobSatisfaction', 'OverTime', 'EmpRelationshipSatisfaction','EmpWorkLifeBalance','Attrition']
df_selected = df[selected_columns]
df_selected.to_excel('processed_data.xlsx', index=False)

### Perform column extraction on the dataset.

In [80]:
missing_values = df_selected.isnull().sum()
print(missing_values)

EmpNumber                      0
Age                            0
Gender                         0
EducationBackground            0
MaritalStatus                  0
EmpDepartment                  0
BusinessTravelFrequency        0
EmpEducationLevel              0
EmpEnvironmentSatisfaction     0
EmpHourlyRate                  0
EmpJobInvolvement              0
EmpJobLevel                    0
EmpJobSatisfaction             0
OverTime                       0
EmpRelationshipSatisfaction    0
EmpWorkLifeBalance             0
Attrition                      0
dtype: int64


### Check for missing values in each column.

In [81]:
# Define the names of the competence rating columns.
comp_columns = ['EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpHourlyRate', 'EmpJobInvolvement',  'EmpJobLevel', 'EmpJobSatisfaction', 'OverTime', 'EmpRelationshipSatisfaction','EmpWorkLifeBalance','Attrition']

# Iterate through these columns to check for anomalies.
for column in comp_columns:
    if column in df_selected.columns: 
        outliers = df_selected[~df_selected[column].isin([1, 2, 3, 4])]
        if not outliers.empty:
            print(f"Anomalies in column '{column}':")
            print(outliers)

Anomalies in column 'EmpEducationLevel':
     EmpNumber  Age  Gender EducationBackground MaritalStatus  \
12    E1001025   30    Male           Marketing      Divorced   
24    E1001053   34  Female       Life Sciences      Divorced   
140   E1001263   39    Male       Life Sciences        Single   
198   E1001351   36  Female             Medical        Single   
209   E1001372   27    Male       Life Sciences       Married   
284   E1001509   42    Male             Medical       Married   
290   E1001520   42  Female             Medical       Married   
392   E1001708   31  Female       Life Sciences        Single   
403   E1001726   41    Male             Medical       Married   
407   E1001737   31  Female       Life Sciences        Single   
421   E1001759   44  Female           Marketing       Married   
441   E1001793   33    Male       Life Sciences        Single   
450   E1001817   50    Male             Medical        Single   
471   E1001850   53  Female    Technical Degree  

### Check for Anomalies

In [82]:
columns_to_process = ['EmpEducationLevel' , 'EmpJobLevel']
for column in columns_to_process:
    df_selected[column] = df_selected[column].apply(lambda x: max(1, min(x, 4)))
for column in columns_to_process:
    if df_selected[column].min() < 1 or df_selected[column].max() > 4:
        print(f"Column {column} still has anomalies")
    else:
        print(f"Column {column} has no anomalies")

Column EmpEducationLevel has no anomalies
Column EmpJobLevel has no anomalies


C:\Users\Crane\AppData\Local\Temp\ipykernel_9728\2361439922.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected[column] = df_selected[column].apply(lambda x: max(1, min(x, 4)))
C:\Users\Crane\AppData\Local\Temp\ipykernel_9728\2361439922.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected[column] = df_selected[column].apply(lambda x: max(1, min(x, 4)))


### Normalize data with minor differences.

In [83]:
# Define the names of the competence rating columns.
comp_columns = ['EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpHourlyRate', 'EmpJobInvolvement',  'EmpJobLevel', 'EmpJobSatisfaction', 'OverTime', 'EmpRelationshipSatisfaction','EmpWorkLifeBalance','Attrition']

# Iterate through these columns to check for anomalies.
for column in comp_columns:
    outliers = df_selected[~df_selected[column].isin([1, 2, 3, 4])]
    if not outliers.empty:
        print(f"Anomalies in column '{column}':")
        print(outliers)

### No more Anomalies

In [84]:
# Define a mapping of original column names to new column names.
column_mapping = {
    'EmpEducationLevel': 'Education Level',
    'EmpEnvironmentSatisfaction': 'Leadership',
    'EmpHourlyRate': 'Communication Skills',
    'EmpJobInvolvement': 'Self-Motivation',
    'EmpJobLevel': 'Problem-Solving Skills',
    'EmpJobSatisfaction': 'Technical Proficiency',
    'OverTime': 'Emotional Intelligence',
    'EmpRelationshipSatisfaction': 'Creativity',
    'EmpWorkLifeBalance': 'Teamwork',
    'Attrition': 'Decision-Making'
}

# Update the column names in the comp_columns list.
comp_columns = [column_mapping[col] if col in column_mapping else col for col in comp_columns]

df_selected.rename(columns=column_mapping, inplace=True)

print(df_selected.columns)



Index(['EmpNumber', 'Age', 'Gender', 'EducationBackground', 'MaritalStatus',
       'EmpDepartment', 'BusinessTravelFrequency', 'Education Level',
       'Leadership', 'Communication Skills', 'Self-Motivation',
       'Problem-Solving Skills', 'Technical Proficiency',
       'Emotional Intelligence', 'Creativity', 'Teamwork', 'Decision-Making'],
      dtype='object')


C:\Users\Crane\AppData\Local\Temp\ipykernel_9728\3415931857.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected.rename(columns=column_mapping, inplace=True)


### Modify the existing competences to match the target competences.

In [85]:
# Replace 'Human Resources' and 'Finance' in the 'EmpDepartment' column with 'Administration'.
df_selected['EmpDepartment'].replace(['Human Resources', 'Finance'], 'Administration', inplace=True)

print(df_selected['EmpDepartment'].unique())

['Sales' 'Administration' 'Development' 'Data Science'
 'Research & Development']


C:\Users\Crane\AppData\Local\Temp\ipykernel_9728\3960424435.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['EmpDepartment'].replace(['Human Resources', 'Finance'], 'Administration', inplace=True)


### Merge 'Human Resources' and 'Finance' departments into 'Administration'.

# Feature Engineering
By applying different weight ratios for various departments, employees are scored accordingly. This process distinguishes two category labels: 'Suitable' and 'Not Suitable' for the position

In [86]:
weights = {
    'Administration': {
        'Leadership': 0.25, 'Communication Skills': 0.20, 'Emotional Intelligence': 0.20, 'Decision-Making': 0.15, 'Teamwork': 0.10,
        'Education Level': 0.05, 'Problem-Solving Skills': 0.05, 'Creativity': 0, 'Technical Proficiency': 0, 'Self-Motivation': 0
    },
    'Sales': {
        'Communication Skills': 0.25, 'Emotional Intelligence': 0.20, 'Self-Motivation': 0.20, 'Creativity': 0.15, 'Teamwork': 0.10,
        'Education Level': 0.05, 'Problem-Solving Skills': 0.05, 'Leadership': 0, 'Technical Proficiency': 0, 'Decision-Making': 0
    },
    'Development': {
        'Technical Proficiency': 0.25, 'Problem-Solving Skills': 0.20, 'Creativity': 0.15, 'Teamwork': 0.10, 'Decision-Making': 0.15,
        'Education Level': 0.05, 'Emotional Intelligence': 0.05, 'Leadership': 0, 'Communication Skills': 0, 'Self-Motivation': 0.05
    },
    'Data Science': {
        'Technical Proficiency': 0.25, 'Problem-Solving Skills': 0.20, 'Creativity': 0.20, 'Education Level': 0.15, 'Decision-Making': 0.10,
        'Teamwork': 0.05, 'Emotional Intelligence': 0.05, 'Leadership': 0, 'Communication Skills': 0, 'Self-Motivation': 0
    },
    'Research & Development': {
        'Technical Proficiency': 0.25, 'Problem-Solving Skills': 0.20, 'Creativity': 0.20, 'Teamwork': 0.15, 'Education Level': 0.10,
        'Decision-Making': 0.05, 'Emotional Intelligence': 0.05, 'Leadership': 0, 'Communication Skills': 0, 'Self-Motivation': 0
    }
}

for department, dept_weights in weights.items():
    df_selected[f'{department}_Score'] = sum(df_selected[comp] * weight for comp, weight in dept_weights.items() if comp in df_selected.columns)
    
threshold = 2  
for department in weights.keys():
    df_selected[f'Suitable_for_{department}'] = df_selected[f'{department}_Score'].apply(lambda x: 'Suit' if x >= threshold else 'NoSuit')

df_selected['Suit_Current_Department'] = df_selected.apply(lambda row: row[f'Suitable_for_{row["EmpDepartment"]}'], axis=1)




C:\Users\Crane\AppData\Local\Temp\ipykernel_9728\2417720358.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected[f'{department}_Score'] = sum(df_selected[comp] * weight for comp, weight in dept_weights.items() if comp in df_selected.columns)
C:\Users\Crane\AppData\Local\Temp\ipykernel_9728\2417720358.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected[f'{department}_Score'] = sum(df_selected[comp] * weight for comp, weight in dept_weights.items() if comp in df_selected.columns)
C:\U

### Weight distribution, output scores for each department.‘Suit’ = 1 、‘Not Suit’ = 0. 

In [87]:
df_selected = pd.get_dummies(df_selected, columns=['Gender', 'EducationBackground', 'MaritalStatus', 'EmpDepartment', 'BusinessTravelFrequency'])

One-hot encoding creates new binary columns

In [88]:
print(df_selected.dtypes)


EmpNumber                                     object
Age                                            int64
Education Level                                int64
Leadership                                     int64
Communication Skills                           int64
Self-Motivation                                int64
Problem-Solving Skills                         int64
Technical Proficiency                          int64
Emotional Intelligence                         int64
Creativity                                     int64
Teamwork                                       int64
Decision-Making                                int64
Administration_Score                         float64
Sales_Score                                  float64
Development_Score                            float64
Data Science_Score                           float64
Research & Development_Score                 float64
Suitable_for_Administration                   object
Suitable_for_Sales                            

In [89]:
columns_to_convert = [
    'Gender_Female', 'Gender_Male',
    'EducationBackground_Human Resources', 'EducationBackground_Life Sciences',
    'EducationBackground_Marketing', 'EducationBackground_Medical',
    'EducationBackground_Other', 'EducationBackground_Technical Degree',
    'MaritalStatus_Divorced', 'MaritalStatus_Married', 'MaritalStatus_Single',
    'EmpDepartment_Administration', 'EmpDepartment_Data Science',
    'EmpDepartment_Development', 'EmpDepartment_Research & Development',
    'EmpDepartment_Sales',
    'BusinessTravelFrequency_Non-Travel',
    'BusinessTravelFrequency_Travel_Frequently',
    'BusinessTravelFrequency_Travel_Rarely'
]
df_selected[columns_to_convert] = df_selected[columns_to_convert].astype(int)


Convert one-hot encoded columns to integer type for machine learning model compatibility.

In [90]:
print(df_selected[columns_to_convert].dtypes)


Gender_Female                                int32
Gender_Male                                  int32
EducationBackground_Human Resources          int32
EducationBackground_Life Sciences            int32
EducationBackground_Marketing                int32
EducationBackground_Medical                  int32
EducationBackground_Other                    int32
EducationBackground_Technical Degree         int32
MaritalStatus_Divorced                       int32
MaritalStatus_Married                        int32
MaritalStatus_Single                         int32
EmpDepartment_Administration                 int32
EmpDepartment_Data Science                   int32
EmpDepartment_Development                    int32
EmpDepartment_Research & Development         int32
EmpDepartment_Sales                          int32
BusinessTravelFrequency_Non-Travel           int32
BusinessTravelFrequency_Travel_Frequently    int32
BusinessTravelFrequency_Travel_Rarely        int32
dtype: object


In [91]:
df_selected.drop('EmpNumber', axis=1, inplace=True)

Remove “EmpNumber”

In [92]:
print(df_selected.dtypes)

Age                                            int64
Education Level                                int64
Leadership                                     int64
Communication Skills                           int64
Self-Motivation                                int64
Problem-Solving Skills                         int64
Technical Proficiency                          int64
Emotional Intelligence                         int64
Creativity                                     int64
Teamwork                                       int64
Decision-Making                                int64
Administration_Score                         float64
Sales_Score                                  float64
Development_Score                            float64
Data Science_Score                           float64
Research & Development_Score                 float64
Suitable_for_Administration                   object
Suitable_for_Sales                            object
Suitable_for_Development                      

Check the types

In [75]:
# Output the modified DataFrame to an Excel file.
df_selected.to_excel('modified_data.xlsx', index=False)

### Print the prototype of the data.


In [94]:
suit_mapping = {'Suit': 1, 'NoSuit': 0}

columns_to_convert = [
    'Suitable_for_Administration', 
    'Suitable_for_Sales', 
    'Suitable_for_Development', 
    'Suitable_for_Data Science', 
    'Suitable_for_Research & Development', 
    'Suit_Current_Department'
]

for col in columns_to_convert:
    df_selected[col] = df_selected[col].replace(suit_mapping)


In [95]:
print(df_selected.dtypes)

Age                                            int64
Education Level                                int64
Leadership                                     int64
Communication Skills                           int64
Self-Motivation                                int64
Problem-Solving Skills                         int64
Technical Proficiency                          int64
Emotional Intelligence                         int64
Creativity                                     int64
Teamwork                                       int64
Decision-Making                                int64
Administration_Score                         float64
Sales_Score                                  float64
Development_Score                            float64
Data Science_Score                           float64
Research & Development_Score                 float64
Suitable_for_Administration                    int64
Suitable_for_Sales                             int64
Suitable_for_Development                      

By encoding, a dataset consisting solely of numerical values is achieved, meeting the requirements of a machine learning dataset

In [96]:
from sklearn.model_selection import train_test_split

X = df_selected.drop('Suit_Current_Department', axis=1)  # Features
y = df_selected['Suit_Current_Department']               # Labels

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


### Split the dataset into training and test sets

In [97]:
print(X_train.dtypes)


Age                                            int64
Education Level                                int64
Leadership                                     int64
Communication Skills                           int64
Self-Motivation                                int64
Problem-Solving Skills                         int64
Technical Proficiency                          int64
Emotional Intelligence                         int64
Creativity                                     int64
Teamwork                                       int64
Decision-Making                                int64
Administration_Score                         float64
Sales_Score                                  float64
Development_Score                            float64
Data Science_Score                           float64
Research & Development_Score                 float64
Suitable_for_Administration                    int64
Suitable_for_Sales                             int64
Suitable_for_Development                      

### Since the "EmpNumber" column is displayed as an "object" type and does not significantly impact the data, it can therefore be removed

# Machine Learning Model Testing
## 1.Random Forest Classifier

In [98]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier()
model.fit(X_train, y_train)

from sklearn.metrics import accuracy_score, classification_report

y_pred = model.predict(X_test)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


0.9861111111111112
              precision    recall  f1-score   support

           0       0.90      0.97      0.93        36
           1       1.00      0.99      0.99       324

    accuracy                           0.99       360
   macro avg       0.95      0.98      0.96       360
weighted avg       0.99      0.99      0.99       360


### This model performs exceptionally well in predicting category 1, with both high precision and recall.
### The prediction for category 0 is also quite accurate, though slightly lower compared to category 1.
### The overall accuracy is very high, indicating that the model makes correct predictions for the majority of samples.

## 2.Support Vector Machine, SVM

In [99]:
from sklearn.svm import SVC

model = SVC()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


0.9
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        36
           1       0.90      1.00      0.95       324

    accuracy                           0.90       360
   macro avg       0.45      0.50      0.47       360
weighted avg       0.81      0.90      0.85       360


D:\Software\Anaconda\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Software\Anaconda\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Software\Anaconda\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#### for this dataset, using SVM might encounter issues such as data imbalance or parameter mismatch

## 3.Decision Tree

In [100]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


0.975
              precision    recall  f1-score   support

           0       0.91      0.83      0.87        36
           1       0.98      0.99      0.99       324

    accuracy                           0.97       360
   macro avg       0.95      0.91      0.93       360
weighted avg       0.97      0.97      0.97       360


## 4.Logistic Regression

In [101]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


0.95
              precision    recall  f1-score   support

           0       0.75      0.75      0.75        36
           1       0.97      0.97      0.97       324

    accuracy                           0.95       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.95      0.95      0.95       360


D:\Software\Anaconda\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#### It is a ConvergenceWarning indicating that the logistic regression model did not converge during training

## 5.K-Nearest Neighbors, KNN

In [102]:
from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


AttributeError: 'Flags' object has no attribute 'c_contiguous'

#### The data type is not suitable for KNN

## conclusion
### Model Testing and Evaluation:

I have tested several popular machine learning models, including Support Vector Machine (SVM), Logistic Regression, Decision Tree, K-Nearest Neighbors (KNN), and Random Forest Classifier (RFC).
During the testing process, I found that models like SVM, Logistic Regression, and KNN encountered errors or warnings in some cases, which might be due to data imbalance, mismatch in model parameters, or the nature of the algorithms themselves.
### Performance of Random Forest Classifier and Decision Tree:

The Random Forest Classifier (RFC) and Decision Tree (DT) performed well in my tests, with no significant errors or issues.
The accuracy of RFC was about 98.61%, while DT achieved an accuracy of 97.5%, both demonstrating high predictive accuracy.
### Future Research Direction:

Based on the current test results and the performance of the models, I plan to continue using Random Forest Classifier and Decision Tree in my future research.
Both models not only offer high accuracy but are also relatively easy to handle with large datasets and have lower requirements for data preprocessing.
### Next Steps:

In future work, I might further fine-tune and optimize the parameters of these two models to enhance their performance.
Additionally, exploring more techniques in feature engineering and model tuning could help improve the models' accuracy and interpretability.
